tools/nightly.py

```diff
@@ -478,6 +478,27 @@ def _link_files(listing, source_dir, target_dir):
         _move_single(src, source_dir, target_dir, os.link, "Linking")
 
 
+def _can_link(platform, source_dir, target_dir):
+    """Determines if we can link from source to target directory"""
+    if platform.startswith("win"):
+        return False
+    src = os.path.join(source_dir, "__nightly_test__")
+    trg = os.path.join(target_dir, "__nightky_test__")
+    try:
+        with open(src, "w"):
+            pass
+        os.link(src, trg)
+        linkable = True
+    except OSError:
+        linkable = False
+    finally:
+        if os.path.isfile(trg):
+            os.remove(trg)
+        if os.path.isfile(src):
+            os.remove(src)
+    return linkable
```

Comment:
Hmm, wouldn't using `os.stat` be a much simpler solution?
```suggestion
    try:
        # Hard linking is possible between two folders on the same device
        return os.stat(source_dir).st_dev == os.stat(target_dir).st_devsrc = os.path.join(source_dir, "__nightly_test__")
     except OSError:
         # hard-linking between non-existant/non-accessible directories is not possible
         return False
```

Reply:
Sorry for taking a long time to get back to this. I don't believe that using stat to test the device is semantically correct here. We want to test if we can link, so the the way to do that is by trying to link a temporary file. The implementation details about which devices allow linking to which other devices isn't a strong enough guarantee that linking can happen, though it may be necessary based on the system

Reply:
Good point. On the other hand, existing implementation does not, strictly speaking, answer the question whether files between two folders can be linked, because it will return false if user is not allowed to create files in source folder or if `__nightly_test__` already exists in source folder or if `__nightky_test__` exists in destination folder.
Considering all that, I think ideal approach would be to just try hard_link and if it fails to copy as suggested in https://github.com/pytorch/pytorch/pull/43771/files#r479539719